salt/transport/tcp.py

```diff
@@ -262,17 +371,95 @@ def _decode_messages(self, messages):
+            await asyncio.sleep(0.001)
+        if timeout == 0:
+            for msg in self.unpacker:
+                framed_msg = salt.transport.frame.decode_embedded_strs(msg)
+                return framed_msg["body"]
+            try:
+                events, _, _ = select.select([self._stream.socket], [], [], 0)
+            except TimeoutError:
+                events = []
+            if events:
+                while not self._closing:
+                    async with self._read_in_progress:
+                        try:
+                            byts = await self._stream.read_bytes(4096, partial=True)
+                        except tornado.iostream.StreamClosedError:
+                            log.trace("Stream closed, reconnecting.")
+                            stream = self._stream
+                            self._stream = None
+                            stream.close()
+                            if self.disconnect_callback:
+                                self.disconnect_callback()
+                            await self.connect()
+                            return
+                        self.unpacker.feed(byts)
+                        for msg in self.unpacker:
+                            framed_msg = salt.transport.frame.decode_embedded_strs(msg)
+                            return framed_msg["body"]
+        elif timeout:
+            try:
+                return await asyncio.wait_for(self.recv(), timeout=timeout)
+            except (
+                TimeoutError,
+                asyncio.exceptions.TimeoutError,
+                asyncio.exceptions.CancelledError,
+            ):
+                self.close()
+                await self.connect()
+                return
+        else:
+            for msg in self.unpacker:
+                framed_msg = salt.transport.frame.decode_embedded_strs(msg)
+                return framed_msg["body"]
+            while not self._closing:
+                async with self._read_in_progress:
+                    try:
+                        byts = await self._stream.read_bytes(4096, partial=True)
+                    except tornado.iostream.StreamClosedError:
+                        log.trace("Stream closed, reconnecting.")
+                        stream = self._stream
+                        self._stream = None
+                        stream.close()
+                        if self.disconnect_callback:
+                            self.disconnect_callback()
+                        await self.connect()
+                        log.debug("Re-connected - continue")
+                        continue
+                    self.unpacker.feed(byts)
+                    for msg in self.unpacker:
+                        framed_msg = salt.transport.frame.decode_embedded_strs(msg)
+                        return framed_msg["body"]
+
+    async def on_recv_handler(self, callback):
+        while not self._stream:
+            await asyncio.sleep(0.003)
```

Comment:
curious as to '0.003', magic number, what give that value significance , and perhaps a global name definition would have helped.

Reply:
Added a comment.